pineA64-modbus-rtu-uart-rs485/vav_thermostat_device_data_read_normal_program/uart_protocol_config_step_pineA64.txt


							uart pins enable for pineA64 board steps




zedbee@pine64:~/hat_modbus1$ =======> ls /dev/ttyS*

/dev/ttyS0  /dev/ttyS2  /dev/ttyS3  /dev/ttyS4  /dev/ttyS5


zedbee@pine64:~/hat_modbus1$ =======> sudo cat /boot/armbianEnv.txt 
verbosity=1
bootlogo=false
console=both
disp_mode=1920x1080p60
overlay_prefix=sun50i-a64
rootdev=UUID=ceaa2402-bfe4-4c0a-a4cb-9492750e40e1
rootfstype=ext4

overlays=uart2 uart3 uart4 i2c0 i2c1 pwm0 pwm1        	<================= add this line all uart enable

param_i2c0_baudrate=100000  # If using bus 1
param_i2c1_baudrate=100000  # If using bus 1
usbstoragequirks=0x2537:0x1066:u,0x2537:0x1068:u



zedbee@pine64:~/hat_modbus1$ =======> sudo reboot



zedbee@pine64:~/hat_modbus1$ =======> sudo cat /sys/kernel/debug/gpio | grep -i pd4


zedbee@pine64:~/hat_modbus1$ =======> sudo cat /sys/kernel/debug/gpio

gpiochip0: GPIOs 0-255, parent: platform/1c20800.pinctrl, 1c20800.pinctrl:
 gpio-166 (                    |cd                  ) in  lo ACTIVE LOW

gpiochip1: GPIOs 352-383, parent: platform/1f02c00.pinctrl, 1f02c00.pinctrl:
 gpio-354 (                    |reset               ) out lo ACTIVE LOW
 gpio-357 (                    |device-wake         ) out hi 

gpiochip2: GPIOs 512-513, parent: platform/axp20x-gpio, axp20x-gpio, can sleep:



zedbee@pine64:~/hat_modbus1$ =======> sudo cat /sys/kernel/debug/gpio | grep -i pd4


zedbee@pine64:~/hat_modbus1$ =======> sudo gpiodetect 

gpiochip0 [1c20800.pinctrl] (256 lines)
gpiochip1 [1f02c00.pinctrl] (32 lines)
gpiochip2 [axp20x-gpio] (2 lines)


zedbee@pine64:~/hat_modbus1$ =======> sudo gpioinfo gpiochip0

gpiochip0 - 256 lines:
	line   0:      unnamed       unused   input  active-high 
	line   1:      unnamed       unused   input  active-high 
	line   2:      unnamed       unused   input  active-high 
	line   3:      unnamed       unused   input  active-high 
	line   4:      unnamed       unused   input  active-high 
	line   5:      unnamed       unused   input  active-high 
	line   6:      unnamed       unused   input  active-high 
	line   7:      unnamed       unused   input  active-high 
	line   8:      unnamed       unused   input  active-high 
	line   9:      unnamed       unused   input  active-high 
	line  10:      unnamed       unused   input  active-high 
	line  11:      unnamed       unused   input  active-high 
	line  12:      unnamed       unused   input  active-high 
	line  13:      unnamed       unused   input  active-high 
	line  14:      unnamed       unused   input  active-high 
	line  15:      unnamed       unused   input  active-high 
	line  16:      unnamed       unused   input  active-high 
	line  17:      unnamed       unused   input  active-high 
	line  18:      unnamed       unused   input  active-high 
	line  19:      unnamed       unused   input  active-high 
	line  20:      unnamed       unused   input  active-high 
	line  21:      unnamed       unused   input  active-high 
	line  22:      unnamed       unused   input  active-high 
	line  23:      unnamed       unused   input  active-high 
	line  24:      unnamed       unused   input  active-high 
	line  25:      unnamed       unused   input  active-high 
	line  26:      unnamed       unused   input  active-high 
	line  27:      unnamed       unused   input  active-high 
	line  28:      unnamed       unused   input  active-high 
	line  29:      unnamed       unused   input  active-high 
	line  30:      unnamed       unused   input  active-high 
	line  31:      unnamed       unused   input  active-high 
	line  32:      unnamed       unused   input  active-high 
	line  33:      unnamed       unused   input  active-high 
	line  34:      unnamed       unused   input  active-high 
	line  35:      unnamed       unused   input  active-high 
	line  36:      unnamed       unused   input  active-high 
	line  37:      unnamed       unused   input  active-high 
	line  38:      unnamed       unused   input  active-high 
	line  39:      unnamed       unused   input  active-high 
	line  40:      unnamed       unused   input  active-high 
	line  41:      unnamed       unused   input  active-high 
	line  42:      unnamed       unused   input  active-high 
	line  43:      unnamed       unused   input  active-high 
	line  44:      unnamed       unused   input  active-high 
	line  45:      unnamed       unused   input  active-high 
	line  46:      unnamed       unused   input  active-high 
	line  47:      unnamed       unused   input  active-high 
	line  48:      unnamed       unused   input  active-high 
	line  49:      unnamed       unused   input  active-high 
	line  50:      unnamed       unused   input  active-high 
	line  51:      unnamed       unused   input  active-high 
	line  52:      unnamed       unused   input  active-high 
	line  53:      unnamed       unused   input  active-high 
	line  54:      unnamed       unused   input  active-high 
	line  55:      unnamed       unused   input  active-high 
	line  56:      unnamed       unused   input  active-high 
	line  57:      unnamed       unused   input  active-high 
	line  58:      unnamed       unused   input  active-high 
	line  59:      unnamed       unused   input  active-high 
	line  60:      unnamed       unused   input  active-high 
	line  61:      unnamed       unused   input  active-high 
	line  62:      unnamed       unused   input  active-high 
	line  63:      unnamed       unused   input  active-high 
	line  64:      unnamed       unused   input  active-high 
	line  65:      unnamed       unused   input  active-high 
	line  66:      unnamed       unused   input  active-high 
	line  67:      unnamed       unused   input  active-high 
	line  68:      unnamed       unused   input  active-high 
	line  69:      unnamed       unused   input  active-high 
	line  70:      unnamed       unused   input  active-high 
	line  71:      unnamed       unused   input  active-high 
	line  72:      unnamed       unused   input  active-high 
	line  73:      unnamed       unused   input  active-high 
	line  74:      unnamed       unused   input  active-high 
	line  75:      unnamed       unused   input  active-high 
	line  76:      unnamed       unused   input  active-high 
	line  77:      unnamed       unused   input  active-high 
	line  78:      unnamed       unused   input  active-high 
	line  79:      unnamed       unused   input  active-high 
	line  80:      unnamed       unused   input  active-high 
	line  81:      unnamed       unused   input  active-high 
	line  82:      unnamed       unused   input  active-high 
	line  83:      unnamed       unused   input  active-high 
	line  84:      unnamed       unused   input  active-high 
	line  85:      unnamed       unused   input  active-high 
	line  86:      unnamed       unused   input  active-high 
	line  87:      unnamed       unused   input  active-high 
	line  88:      unnamed       unused   input  active-high 
	line  89:      unnamed       unused   input  active-high 
	line  90:      unnamed       unused   input  active-high 
	line  91:      unnamed       unused   input  active-high 
	line  92:      unnamed       unused   input  active-high 
	line  93:      unnamed       unused   input  active-high 
	line  94:      unnamed       unused   input  active-high 
	line  95:      unnamed       unused   input  active-high 
	line  96:      unnamed       unused   input  active-high 
	line  97:      unnamed       unused   input  active-high 
	line  98:      unnamed       unused   input  active-high 
	line  99:      unnamed       unused   input  active-high 

	line 100:      unnamed       unused  output  active-high    <================  this for DE3 pin for gpiochip0

	line 101:      unnamed       unused   input  active-high 

	line 102:      unnamed       unused  output  active-high    <================  this for DE2 pin for gpiochip0


	line 103:      unnamed       unused   input  active-high 
	line 104:      unnamed       unused   input  active-high 
	line 105:      unnamed       unused   input  active-high 
	line 106:      unnamed       unused   input  active-high 
	line 107:      unnamed       unused   input  active-high 
	line 108:      unnamed       unused   input  active-high 
	line 109:      unnamed       unused   input  active-high 
	line 110:      unnamed       unused   input  active-high 
	line 111:      unnamed       unused   input  active-high 
	line 112:      unnamed       unused   input  active-high 
	line 113:      unnamed       unused   input  active-high 
	line 114:      unnamed       unused   input  active-high 
	line 115:      unnamed       unused   input  active-high 
	line 116:      unnamed       unused   input  active-high 
	line 117:      unnamed       unused   input  active-high 
	line 118:      unnamed       unused   input  active-high 
	line 119:      unnamed       unused   input  active-high 
	line 120:      unnamed       unused   input  active-high 
	line 121:      unnamed       unused   input  active-high 
	line 122:      unnamed       unused   input  active-high 
	line 123:      unnamed       unused   input  active-high 
	line 124:      unnamed       unused   input  active-high 
	line 125:      unnamed       unused   input  active-high 
	line 126:      unnamed       unused   input  active-high 
	line 127:      unnamed       unused   input  active-high 
	line 128:      unnamed       unused   input  active-high 
	line 129:      unnamed       unused   input  active-high 
	line 130:      unnamed       unused   input  active-high 
	line 131:      unnamed       unused   input  active-high 
	line 132:      unnamed       unused   input  active-high 
	line 133:      unnamed       unused   input  active-high 
	line 134:      unnamed       unused   input  active-high 
	line 135:      unnamed       unused   input  active-high 
	line 136:      unnamed       unused   input  active-high 
	line 137:      unnamed       unused   input  active-high 
	line 138:      unnamed       unused   input  active-high 
	line 139:      unnamed       unused   input  active-high 
	line 140:      unnamed       unused   input  active-high 
	line 141:      unnamed       unused   input  active-high 
	line 142:      unnamed       unused   input  active-high 
	line 143:      unnamed       unused   input  active-high 
	line 144:      unnamed       unused   input  active-high 
	line 145:      unnamed       unused   input  active-high 
	line 146:      unnamed       unused   input  active-high 
	line 147:      unnamed       unused   input  active-high 
	line 148:      unnamed       unused   input  active-high 
	line 149:      unnamed       unused   input  active-high 
	line 150:      unnamed       unused   input  active-high 
	line 151:      unnamed       unused   input  active-high 
	line 152:      unnamed       unused   input  active-high 
	line 153:      unnamed       unused   input  active-high 
	line 154:      unnamed       unused   input  active-high 
	line 155:      unnamed       unused   input  active-high 
	line 156:      unnamed       unused   input  active-high 
	line 157:      unnamed       unused   input  active-high 
	line 158:      unnamed       unused   input  active-high 
	line 159:      unnamed       unused   input  active-high 
	line 160:      unnamed       unused   input  active-high 
	line 161:      unnamed       unused   input  active-high 
	line 162:      unnamed       unused   input  active-high 
	line 163:      unnamed       unused   input  active-high 
	line 164:      unnamed       unused   input  active-high 
	line 165:      unnamed       unused   input  active-high 
	line 166:      unnamed         "cd"   input   active-low [used]
	line 167:      unnamed       unused   input  active-high 
	line 168:      unnamed       unused   input  active-high 
	line 169:      unnamed       unused   input  active-high 
	line 170:      unnamed       unused   input  active-high 
	line 171:      unnamed       unused   input  active-high 
	line 172:      unnamed       unused   input  active-high 
	line 173:      unnamed       unused   input  active-high 
	line 174:      unnamed       unused   input  active-high 
	line 175:      unnamed       unused   input  active-high 
	line 176:      unnamed       unused   input  active-high 
	line 177:      unnamed       unused   input  active-high 
	line 178:      unnamed       unused   input  active-high 
	line 179:      unnamed       unused   input  active-high 
	line 180:      unnamed       unused   input  active-high 
	line 181:      unnamed       unused   input  active-high 
	line 182:      unnamed       unused   input  active-high 
	line 183:      unnamed       unused   input  active-high 
	line 184:      unnamed       unused   input  active-high 
	line 185:      unnamed       unused   input  active-high 
	line 186:      unnamed       unused   input  active-high 
	line 187:      unnamed       unused   input  active-high 
	line 188:      unnamed       unused   input  active-high 
	line 189:      unnamed       unused   input  active-high 
	line 190:      unnamed       unused   input  active-high 
	line 191:      unnamed       unused   input  active-high 
	line 192:      unnamed       unused   input  active-high 
	line 193:      unnamed       unused   input  active-high 
	line 194:      unnamed       unused   input  active-high 
	line 195:      unnamed       unused   input  active-high 
	line 196:      unnamed       unused   input  active-high 
	line 197:      unnamed       unused   input  active-high 
	line 198:      unnamed       unused   input  active-high 
	line 199:      unnamed       unused   input  active-high 
	line 200:      unnamed       unused   input  active-high 
	line 201:      unnamed       unused   input  active-high 
	line 202:      unnamed       unused   input  active-high 
	line 203:      unnamed       unused   input  active-high 
	line 204:      unnamed       unused   input  active-high 
	line 205:      unnamed       unused   input  active-high 
	line 206:      unnamed       unused   input  active-high 
	line 207:      unnamed       unused   input  active-high 
	line 208:      unnamed       unused   input  active-high 
	line 209:      unnamed       unused   input  active-high 
	line 210:      unnamed       unused   input  active-high 
	line 211:      unnamed       unused   input  active-high 
	line 212:      unnamed       unused   input  active-high 
	line 213:      unnamed       unused   input  active-high 
	line 214:      unnamed       unused   input  active-high 
	line 215:      unnamed       unused   input  active-high 
	line 216:      unnamed       unused   input  active-high 
	line 217:      unnamed       unused   input  active-high 
	line 218:      unnamed       unused   input  active-high 
	line 219:      unnamed       unused   input  active-high 
	line 220:      unnamed       unused   input  active-high 
	line 221:      unnamed       unused   input  active-high 
	line 222:      unnamed       unused   input  active-high 
	line 223:      unnamed       unused   input  active-high 
	line 224:      unnamed       unused   input  active-high 
	line 225:      unnamed       unused   input  active-high 
	line 226:      unnamed       unused   input  active-high 
	line 227:      unnamed       unused   input  active-high 
	line 228:      unnamed       unused   input  active-high 
	line 229:      unnamed       unused   input  active-high 
	line 230:      unnamed       unused   input  active-high 
	line 231:      unnamed       unused   input  active-high 
	line 232:      unnamed       unused   input  active-high 
	line 233:      unnamed       unused   input  active-high 
	line 234:      unnamed       unused   input  active-high 
	line 235:      unnamed       unused   input  active-high 
	line 236:      unnamed       unused   input  active-high 
	line 237:      unnamed       unused   input  active-high 
	line 238:      unnamed       unused   input  active-high 
	line 239:      unnamed       unused   input  active-high 
	line 240:      unnamed       unused   input  active-high 
	line 241:      unnamed       unused   input  active-high 
	line 242:      unnamed       unused   input  active-high 
	line 243:      unnamed       unused   input  active-high 
	line 244:      unnamed       unused   input  active-high 
	line 245:      unnamed       unused   input  active-high 
	line 246:      unnamed       unused   input  active-high 
	line 247:      unnamed       unused   input  active-high 
	line 248:      unnamed       unused   input  active-high 
	line 249:      unnamed       unused   input  active-high 
	line 250:      unnamed       unused   input  active-high 
	line 251:      unnamed       unused   input  active-high 
	line 252:      unnamed       unused   input  active-high 
	line 253:      unnamed       unused   input  active-high 
	line 254:      unnamed       unused   input  active-high 
	line 255:      unnamed       unused   input  active-high 


	
zedbee@pine64:~/hat_modbus1$ =======> dmesg | grep tty

[    0.000000] Kernel command line: root=UUID=ceaa2402-bfe4-4c0a-a4cb-9492750e40e1 rootwait rootfstype=ext4 splash=verbose console=ttyS0,115200 console=tty1 consoleblank=0 loglevel=1 ubootpart=aaa8996f-01 usb-storage.quirks=0x2537:0x1066:u,0x2537:0x1068:u   cgroup_enable=memory
[    0.001004] printk: legacy console [tty1] enabled
[    1.750189] printk: legacy console [ttyS0] disabled
[    1.750674] 1c28000.serial: ttyS0 at MMIO 0x1c28000 (irq = 155, base_baud = 1500000) is a 16550A
[    1.750740] printk: legacy console [ttyS0] enabled
[    1.753054] 1c28400.serial: ttyS1 at MMIO 0x1c28400 (irq = 156, base_baud = 1500000) is a 16550A
[    1.753298] serial serial0: tty port ttyS1 registered
[    1.754549] 1c28800.serial: ttyS2 at MMIO 0x1c28800 (irq = 157, base_baud = 1500000) is a 16550A
[    1.756332] 1c28c00.serial: ttyS3 at MMIO 0x1c28c00 (irq = 158, base_baud = 1500000) is a 16550A
[    1.757834] 1c29000.serial: ttyS4 at MMIO 0x1c29000 (irq = 159, base_baud = 1500000) is a 16550A
[    7.557368] systemd[1]: Created slice system-getty.slice - Slice /system/getty.
[    7.565792] systemd[1]: Created slice system-serial\x2dgetty.slice - Slice /system/serial-getty.
[    7.570112] systemd[1]: Expecting device dev-ttyS0.device - /dev/ttyS0...
[    8.954225] systemd[1]: Found device dev-ttyS0.device - /dev/ttyS0.
[    9.885976] ch341-uart ttyUSB0: break control not supported, using simulated break
[    9.890024] usb 4-1: ch341-uart converter now attached to ttyUSB0
[  510.192299] ch341-uart ttyUSB0: ch341-uart converter now disconnected from ttyUSB0



zedbee@pine64:~/hat_modbus1$ =======> dmesg | grep serial

[    1.750674] 1c28000.serial: ttyS0 at MMIO 0x1c28000 (irq = 155, base_baud = 1500000) is a 16550A
[    1.753054] 1c28400.serial: ttyS1 at MMIO 0x1c28400 (irq = 156, base_baud = 1500000) is a 16550A
[    1.753298] serial serial0: tty port ttyS1 registered
[    1.754549] 1c28800.serial: ttyS2 at MMIO 0x1c28800 (irq = 157, base_baud = 1500000) is a 16550A
[    1.756332] 1c28c00.serial: ttyS3 at MMIO 0x1c28c00 (irq = 158, base_baud = 1500000) is a 16550A
[    1.757834] 1c29000.serial: ttyS4 at MMIO 0x1c29000 (irq = 159, base_baud = 1500000) is a 16550A
[    3.122589] dw-apb-uart 1c28000.serial: forbid DMA for kernel console
[    7.565792] systemd[1]: Created slice system-serial\x2dgetty.slice - Slice /system/serial-getty.
[    9.703119] usbcore: registered new interface driver usbserial_generic
[    9.703184] usbserial: USB Serial support registered for generic
[    9.875065] usbserial: USB Serial support registered for ch341-uart
[   10.852442] dw-apb-uart 1c28400.serial: failed to request DMA
[  201.754232] dw-apb-uart 1c29000.serial: failed to request DMA
[  309.505326] dw-apb-uart 1c28c00.serial: failed to request DMA



pineA64-modbus-rtu-uart-rs485/vav_thermostat_read_data_sendto_mqtt_bms_cloude_program/config.h
#ifndef CONFIG_H
#define CONFIG_H


#include <modbus/modbus.h>
#include <unistd.h>

#define DEVICE		"/dev/ttyS3" 	//  this for serial modbus rtu usb connect "/dev/ttyUSB0",
					//  uart 2 this for serial modbus rtu with pine hat board // "/dev/ttyS2"  // 8 --> UART2_TXD GPIO 14 PB0, 10 --> UART_RXD GPIO 15 PB1 ---> Pi-2 Connector
					//  uart 3 this for serial modbus rtu with pine hat board // "/dev/ttyS3"  // 24 --> UART3_TXD PD0, 23 --> UART3_RXD PD1  ---> Euler "e" Connector
					//  uart 4 this for serial modbus rtu with pine hat board // "/dev/ttyS4"  // 19 --> UART4_TXD PD2, 21 --> UART4_RXD PD3  ---> Euler "e" Connector

#define BAUD_RATE 	19200		//  Baud Rate ---> 19200
#define PARITY		'N'		//  Parity for N ---> None
#define DATA_BITS 	8		//  Data Bits ---> 8
#define STOP_BITS	1		//  Stop Bits ---> 1
#define SLAVE_ID	2  // 12	//  Slave ID ---> 6
#define NUM_REGS	18		//  Number of Register ---> 18

#define DE_GPIO_CHIP 	"/dev/gpiochip0"  // this for gpiochip0 in DE pins enable select this 
#define DE_GPIO_LINE	102 	// this for uart 3 DE2 pin for = PD6  // 102 ---> // GPIO102 = PD6 (RS485 DE/RE pin), ---> Euler "e" Connector  
				// this for uart 4 DE3 pin for = PD4  // 100 ---> // GPIO100 = PD4 (RS485 DE/RE pin), ---> Euler "e" Connector
				// this for uart 2 DE1 pin for = PC15 //  79 ---> // GPIO79 = PC15 (RS485 DE/RE pin)  ---> Pi-2 Connector

#define BUFFER_SIZE1 30000	// Buffer size 30000


// Boodskap / MQTT
#define BROKER_DOMAIN_KEY   "BZHEZISEWY"
#define BROKER_API_KEY      "Yrzp62Nu20H0" // "Yrzp62Nu20H0" ---> this for v4dev in iitmrp    // "UUOuKQvTSUtv" this for zedbee.io in production main
#define BOODSKAP_BROKER     "v4dev.zedbee.in" // tcp://v4dev.zedbee.in:1883 ---> this for v4dev in iitmrp  // "zedbee.io" ---> this for BMS cloude main
#define BOODSKAP_USERNAME   "DEV_" BROKER_DOMAIN_KEY  	//  boodskap username ---> "DEV_BZHEZISEWY"
#define BOODSKAP_PASSWORD   BROKER_API_KEY	 	//  boodskap password ---> "Yrzp62Nu20H0"


#define B_CUSTOMER_ID       "ZBCID00012"  // "ZBCID00030". customer id will be changed
#define B_TENANT_ID         "ZBTID0001"	  // "ZBTID0001". Tenant id not changed
#define B_DEVICE_ID         "ZBDID02BA86068689"	 // ZBDID02BA86068689". device id will be changed

#define BOODSKAP_PUBLISH_TOPIC  "/BZHEZISEWY/device/ZBDID02BA86068689/msgs/Gateway/1/101"// "/" BROKER_DOMAIN_KEY "/device" B_DEVICE_ID "/msgs/gateway/1/101"
#define BOODSKAP_SUBSCRIBE_TOPIC "/BZHEZISEWY/device/ZBDID02BA86068689/cmds"  // "/" BROKER_DOMAIN_KEY "/device/" B_DEVICE_ID "/cmds"
#define BOODSKAP_ACK_TOPIC       "/BZHEZISEWY/device/ZBDID02BA86068689/msgs/Gateway/1/103" // "/" BROKER_DOMAIN_KEY "/device/" B_DEVICE_ID "/msgs/gateway/1/103"

#define CLIENTID            "DEV_" B_DEVICE_ID   // "DEV_ZBDID02BA86068689"  device id will be changed
#define PUBLISH_INTERVAL    180


#endif // CONFIG_H
